File: lib/ts/src/internal/GetApiDefinitionsWithExamples.ts
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

import { RandomGenerator } from '../RandomGenerator.js';
import { TelepactSchema } from '../TelepactSchema.js';
import { encodeBase64 } from './binary/Base64Util.js';
import { GenerateContext } from './generation/GenerateContext.js';
import { TFieldDeclaration } from './types/TFieldDeclaration.js';
import { TType } from './types/TType.js';

const EXAMPLE_COLLECTION_LENGTH = 2;

export function getApiDefinitionsWithExamples(telepactSchema: TelepactSchema, includeInternal: boolean): any[] {
    const definitions = includeInternal ? telepactSchema.full : telepactSchema.original;
    const defaultFnScope = getDefaultFnScope(telepactSchema.parsed);

    return definitions.map((definition) =>
        addExamplesToDefinition(definition as Record<string, any>, telepactSchema, defaultFnScope)
    );
}

function addExamplesToDefinition(
    definition: Record<string, any>,
    telepactSchema: TelepactSchema,
    defaultFnScope: string,
): Record<string, any> {
    const schemaKey = getSchemaKey(definition);
    const clonedDefinition = { ...definition };

    if (schemaKey.startsWith('info.')) {
        clonedDefinition.example = {};
        return clonedDefinition;
    }

    const randomGenerator = new RandomGenerator(EXAMPLE_COLLECTION_LENGTH, EXAMPLE_COLLECTION_LENGTH);

    if (schemaKey.startsWith('fn.')) {
        const ctx = new GenerateContext(true, false, true, schemaKey, randomGenerator);
        clonedDefinition.inputExample = normalizeExampleValue(
            telepactSchema.parsed[schemaKey].generateRandomValue(null, false, [], ctx),
        );
        clonedDefinition.outputExample = normalizeExampleValue(
            telepactSchema.parsed[`${schemaKey}.->`].generateRandomValue(null, false, [], ctx),
        );
        return clonedDefinition;
    }

    if (schemaKey.startsWith('headers.')) {
        const ctx = new GenerateContext(true, false, true, defaultFnScope, randomGenerator);
        clonedDefinition.inputExample = generateHeaderExample(
            definition[schemaKey] as Record<string, any>,
            telepactSchema.parsedRequestHeaders,
            ctx,
        );
        clonedDefinition.outputExample = generateHeaderExample(
            definition['->'] as Record<string, any>,
            telepactSchema.parsedResponseHeaders,
            ctx,
        );
        return clonedDefinition;
    }

    if (schemaKey.startsWith('errors.')) {
        const ctx = new GenerateContext(true, false, true, defaultFnScope, randomGenerator);
        clonedDefinition.example = generateRawUnionExample(definition[schemaKey] as any[], telepactSchema, ctx);
        return clonedDefinition;
    }

    const ctx = new GenerateContext(true, false, true, defaultFnScope, randomGenerator);
    clonedDefinition.example = normalizeExampleValue(
        telepactSchema.parsed[schemaKey].generateRandomValue(null, false, [], ctx),
    );
    return clonedDefinition;
}

function generateHeaderExample(
    headerDefinition: Record<string, any> | undefined,
    parsedHeaders: Record<string, TFieldDeclaration>,
    ctx: GenerateContext,
): Record<string, any> {
    const example: Record<string, any> = {};

    for (const headerName of Object.keys(headerDefinition || {}).sort()) {
        example[headerName] = normalizeExampleValue(
            parsedHeaders[headerName].typeDeclaration.generateRandomValue(null, false, ctx),
        );
    }

    return example;
}

function generateRawUnionExample(unionDefinition: any[], telepactSchema: TelepactSchema, ctx: GenerateContext): Record<string, any> {
    const tags = unionDefinition
        .map((tagDefinition) => {
            const [tagName, tagPayload] = Object.entries(tagDefinition as Record<string, any>)
                .find(([key]) => key !== '///') as [string, Record<string, any>];
            return [tagName, tagPayload] as [string, Record<string, any>];
        })
        .sort(([left], [right]) => left.localeCompare(right));
    const [tagName, tagPayload] = tags[ctx.randomGenerator.nextIntWithCeiling(tags.length)];
    return {
        [tagName]: generateRawStructExample(tagPayload, telepactSchema, ctx),
    };
}

function generateRawStructExample(
    structDefinition: Record<string, any>,
    telepactSchema: TelepactSchema,
    ctx: GenerateContext,
): Record<string, any> {
    const example: Record<string, any> = {};

    for (const fieldName of Object.keys(structDefinition).sort()) {
        const optional = fieldName.endsWith('!');
        if (optional) {
            if (!ctx.includeOptionalFields || (ctx.randomizeOptionalFields && ctx.randomGenerator.nextBoolean())) {
                continue;
            }
        } else if (!ctx.alwaysIncludeRequiredFields && ctx.randomGenerator.nextBoolean()) {
            continue;
        }

        example[fieldName] = generateRawTypeExample(structDefinition[fieldName], telepactSchema, ctx);
    }

    return example;
}

function generateRawTypeExample(typeExpression: any, telepactSchema: TelepactSchema, ctx: GenerateContext): any {
    if (typeof typeExpression === 'string') {
        const nullable = typeExpression.endsWith('?');
        const nonNullableTypeExpression = nullable ? typeExpression.slice(0, -1) : typeExpression;
        if (nullable && ctx.randomGenerator.nextBoolean()) {
            return null;
        }

        switch (nonNullableTypeExpression) {
            case 'boolean':
                return ctx.randomGenerator.nextBoolean();
            case 'integer':
                return ctx.randomGenerator.nextInt();
            case 'number':
                return ctx.randomGenerator.nextDouble();
            case 'string':
                return ctx.randomGenerator.nextString();
            case 'any': {
                const selectType = ctx.randomGenerator.nextIntWithCeiling(3);
                if (selectType === 0) {
                    return ctx.randomGenerator.nextBoolean();
                }
                if (selectType === 1) {
                    return ctx.randomGenerator.nextInt();
                }
                return ctx.randomGenerator.nextString();
            }
            case 'bytes':
                return encodeBase64(ctx.randomGenerator.nextBytes());
            default:
                return normalizeExampleValue(
                    telepactSchema.parsed[nonNullableTypeExpression].generateRandomValue(null, false, [], ctx),
                );
        }
    }

    if (Array.isArray(typeExpression)) {
        const length = ctx.randomGenerator.nextCollectionLength();
        const example = [];
        for (let i = 0; i < length; i += 1) {
            example.push(generateRawTypeExample(typeExpression[0], telepactSchema, ctx));
        }
        return example;
    }

    if (typeof typeExpression === 'object' && typeExpression !== null) {
        const objectDefinition = typeExpression as Record<string, any>;
        const [key] = Object.keys(objectDefinition);
        if (key === 'string') {
            const length = ctx.randomGenerator.nextCollectionLength();
            const example: Record<string, any> = {};
            for (let i = 0; i < length; i += 1) {
                example[ctx.randomGenerator.nextString()] = generateRawTypeExample(objectDefinition[key], telepactSchema, ctx);
            }
            return example;
        }
    }

    return null;
}

function normalizeExampleValue(value: any): any {
    if (value instanceof Uint8Array) {
        return encodeBase64(value);
    }

    if (Array.isArray(value)) {
        return value.map((entry) => normalizeExampleValue(entry));
    }

    if (typeof value === 'object' && value !== null) {
        const normalized: Record<string, any> = {};
        for (const [key, entry] of Object.entries(value)) {
            normalized[key] = normalizeExampleValue(entry);
        }
        return normalized;
    }

    return value;
}

function getDefaultFnScope(parsedTypes: Record<string, TType>): string {
    const nonInternalFunctions = Object.keys(parsedTypes)
        .filter((schemaKey) => schemaKey.startsWith('fn.') && !schemaKey.endsWith('.->') && !schemaKey.endsWith('_'))
        .sort();
    if (nonInternalFunctions.length > 0) {
        return nonInternalFunctions[0];
    }

    const allFunctions = Object.keys(parsedTypes)
        .filter((schemaKey) => schemaKey.startsWith('fn.') && !schemaKey.endsWith('.->'))
        .sort();
    if (allFunctions.length > 0) {
        return allFunctions[0];
    }

    return 'fn.ping_';
}

function getSchemaKey(definition: Record<string, any>): string {
    for (const key of Object.keys(definition)) {
        if (key !== '///' && key !== '->' && key !== '_errors') {
            return key;
        }
    }

    throw new Error(`Schema entry has no schema key: ${JSON.stringify(definition)}`);
}

File: lib/go/internal/GetAPIDefinitionsWithExamples.go
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package internal

import (
	"encoding/base64"
	"encoding/binary"
	"fmt"
	"sort"
	"strings"

	"github.com/telepact/telepact/lib/go/internal/types"
)

const exampleCollectionLength = 2

var exampleRandomWords = []string{
	"alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta", "theta",
	"iota", "kappa", "lambda", "mu", "nu", "xi", "omicron", "pi",
	"rho", "sigma", "tau", "upsilon", "phi", "chi", "psi", "omega",
}

type apiExampleRandomGenerator struct {
	seed                int32
	collectionLengthMin int
	collectionLengthMax int
}

func newAPIExampleRandomGenerator() *apiExampleRandomGenerator {
	random := &apiExampleRandomGenerator{
		collectionLengthMin: exampleCollectionLength,
		collectionLengthMax: exampleCollectionLength,
	}
	random.SetSeed(0)
	return random
}

func (r *apiExampleRandomGenerator) SetSeed(seed int32) {
	if seed == 0 {
		r.seed = 1
		return
	}
	r.seed = seed
}

func (r *apiExampleRandomGenerator) NextInt() int {
	x := r.seed
	x ^= x << 16
	x ^= x >> 11
	x ^= x << 5
	if x == 0 {
		x = 1
	}
	r.seed = x
	return int(x & 0x7fffffff)
}

func (r *apiExampleRandomGenerator) NextIntWithCeiling(ceiling int) int {
	if ceiling == 0 {
		return 0
	}
	return r.NextInt() % ceiling
}

func (r *apiExampleRandomGenerator) NextBoolean() bool {
	return r.NextIntWithCeiling(31) > 15
}

func (r *apiExampleRandomGenerator) NextString() string {
	return exampleRandomWords[r.NextIntWithCeiling(len(exampleRandomWords))]
}

func (r *apiExampleRandomGenerator) NextCollectionLength() int {
	offset := r.collectionLengthMax - r.collectionLengthMin
	if offset <= 0 {
		return r.collectionLengthMin
	}
	return r.NextIntWithCeiling(offset) + r.collectionLengthMin
}

func (r *apiExampleRandomGenerator) NextBytes() []byte {
	bytes := make([]byte, 4)
	binary.BigEndian.PutUint32(bytes, uint32(r.NextInt()))
	return bytes
}

func (r *apiExampleRandomGenerator) NextDouble() float64 {
	return float64(r.NextInt()&0x7fffffff) / float64(0x7fffffff)
}

// GetAPIDefinitionsWithExamples returns schema definitions annotated with deterministic example payloads.
func GetAPIDefinitionsWithExamples(schema SchemaAccessor, includeInternal bool) []any {
	definitions := schema.OriginalDefinitions()
	if includeInternal {
		definitions = schema.FullDefinitions()
	}

	defaultFnScope := getDefaultFnScope(schema.ParsedDefinitions())
	result := make([]any, 0, len(definitions))
	for _, definition := range definitions {
		definitionMap, ok := definition.(map[string]any)
		if !ok {
			result = append(result, definition)
			continue
		}
		result = append(result, addExamplesToDefinition(definitionMap, schema, defaultFnScope))
	}

	return result
}

// GetDefinitionExample returns example payload fields for a single schema definition.
func GetDefinitionExample(schema SchemaAccessor, name string, includeInternal bool) map[string]any {
	if !includeInternal && strings.HasSuffix(name, "_") {
		return map[string]any{}
	}

	definitions := schema.OriginalDefinitions()
	if includeInternal {
		definitions = schema.FullDefinitions()
	}

	for _, definition := range definitions {
		definitionMap := toStringAnyMap(definition)
		if getSchemaKey(definitionMap) != name {
			continue
		}

		exampleDefinition := addExamplesToDefinition(
			definitionMap,
			schema,
			getDefaultFnScope(schema.ParsedDefinitions()),
		)
		result := map[string]any{}
		for _, key := range []string{"example", "inputExample", "outputExample"} {
			if value, ok := exampleDefinition[key]; ok {
				result[key] = value
			}
		}
		return result
	}

	return map[string]any{}
}

func addExamplesToDefinition(
	definition map[string]any,
	schema SchemaAccessor,
	defaultFnScope string,
) map[string]any {
	schemaKey := getSchemaKey(definition)
	clonedDefinition := cloneStringAnyMap(definition)

	if strings.HasPrefix(schemaKey, "info.") {
		clonedDefinition["example"] = map[string]any{}
		return clonedDefinition
	}

	randomGenerator := newAPIExampleRandomGenerator()

	if strings.HasPrefix(schemaKey, "fn.") {
		ctx := types.NewGenerateContext(true, false, true, schemaKey, randomGenerator)
		clonedDefinition["inputExample"] = normalizeExampleValue(
			schema.ParsedDefinitions()[schemaKey].GenerateRandomValue(nil, false, nil, ctx),
		)
		clonedDefinition["outputExample"] = normalizeExampleValue(
			schema.ParsedDefinitions()[schemaKey+".->"].GenerateRandomValue(nil, false, nil, ctx),
		)
		return clonedDefinition
	}

	if strings.HasPrefix(schemaKey, "headers.") {
		ctx := types.NewGenerateContext(true, false, true, defaultFnScope, randomGenerator)
		clonedDefinition["inputExample"] = generateHeaderExample(
			toStringAnyMap(definition[schemaKey]),
			schema.RequestHeaderDeclarations(),
			ctx,
		)
		clonedDefinition["outputExample"] = generateHeaderExample(
			toStringAnyMap(definition["->"]),
			schema.ResponseHeaderDeclarations(),
			ctx,
		)
		return clonedDefinition
	}

	if strings.HasPrefix(schemaKey, "errors.") {
		ctx := types.NewGenerateContext(true, false, true, defaultFnScope, randomGenerator)
		clonedDefinition["example"] = generateRawUnionExample(
			toAnySlice(definition[schemaKey]),
			schema,
			ctx,
		)
		return clonedDefinition
	}

	ctx := types.NewGenerateContext(true, false, true, defaultFnScope, randomGenerator)
	clonedDefinition["example"] = normalizeExampleValue(
		schema.ParsedDefinitions()[schemaKey].GenerateRandomValue(nil, false, nil, ctx),
	)
	return clonedDefinition
}

func generateHeaderExample(
	headerDefinition map[string]any,
	parsedHeaders map[string]*types.TFieldDeclaration,
	ctx *types.GenerateContext,
) map[string]any {
	if headerDefinition == nil {
		return map[string]any{}
	}

	headerNames := make([]string, 0, len(headerDefinition))
	for headerName := range headerDefinition {
		headerNames = append(headerNames, headerName)
	}
	sort.Strings(headerNames)

	example := make(map[string]any, len(headerNames))
	for _, headerName := range headerNames {
		example[headerName] = normalizeExampleValue(
			parsedHeaders[headerName].TypeDeclaration.GenerateRandomValue(nil, false, ctx),
		)
	}

	return example
}

func generateRawUnionExample(
	unionDefinition []any,
	schema SchemaAccessor,
	ctx *types.GenerateContext,
) map[string]any {
	type tagDefinition struct {
		name    string
		payload map[string]any
	}

	tags := make([]tagDefinition, 0, len(unionDefinition))
	for _, entry := range unionDefinition {
		tagMap := toStringAnyMap(entry)
		for key, value := range tagMap {
			if key == "///" {
				continue
			}
			tags = append(tags, tagDefinition{
				name:    key,
				payload: toStringAnyMap(value),
			})
			break
		}
	}

	sort.Slice(tags, func(i, j int) bool {
		return tags[i].name < tags[j].name
	})

	tag := tags[ctx.RandomGenerator.NextIntWithCeiling(len(tags))]
	return map[string]any{
		tag.name: generateRawStructExample(tag.payload, schema, ctx),
	}
}

func generateRawStructExample(
	structDefinition map[string]any,
	schema SchemaAccessor,
	ctx *types.GenerateContext,
) map[string]any {
	fieldNames := make([]string, 0, len(structDefinition))
	for fieldName := range structDefinition {
		fieldNames = append(fieldNames, fieldName)
	}
	sort.Strings(fieldNames)

	example := make(map[string]any)
	for _, fieldName := range fieldNames {
		optional := strings.HasSuffix(fieldName, "!")
		if optional {
			if !ctx.IncludeOptionalFields || (ctx.RandomizeOptionalFields && ctx.RandomGenerator.NextBoolean()) {
				continue
			}
		} else if !ctx.AlwaysIncludeRequiredFields && ctx.RandomGenerator.NextBoolean() {
			continue
		}

		example[fieldName] = generateRawTypeExample(structDefinition[fieldName], schema, ctx)
	}

	return example
}

func generateRawTypeExample(
	typeExpression any,
	schema SchemaAccessor,
	ctx *types.GenerateContext,
) any {
	switch typed := typeExpression.(type) {
	case string:
		nullable := strings.HasSuffix(typed, "?")
		nonNullableTypeExpression := typed
		if nullable {
			nonNullableTypeExpression = strings.TrimSuffix(typed, "?")
			if ctx.RandomGenerator.NextBoolean() {
				return nil
			}
		}

		switch nonNullableTypeExpression {
		case "boolean":
			return ctx.RandomGenerator.NextBoolean()
		case "integer":
			return ctx.RandomGenerator.NextInt()
		case "number":
			return ctx.RandomGenerator.NextDouble()
		case "string":
			return ctx.RandomGenerator.NextString()
		case "any":
			selectType := ctx.RandomGenerator.NextIntWithCeiling(3)
			switch selectType {
			case 0:
				return ctx.RandomGenerator.NextBoolean()
			case 1:
				return ctx.RandomGenerator.NextInt()
			default:
				return ctx.RandomGenerator.NextString()
			}
		case "bytes":
			return base64.StdEncoding.EncodeToString(ctx.RandomGenerator.NextBytes())
		default:
			return normalizeExampleValue(
				schema.ParsedDefinitions()[nonNullableTypeExpression].GenerateRandomValue(nil, false, nil, ctx),
			)
		}

	case []any:
		length := ctx.RandomGenerator.NextCollectionLength()
		example := make([]any, 0, length)
		for i := 0; i < length; i += 1 {
			example = append(example, generateRawTypeExample(typed[0], schema, ctx))
		}
		return example

	case map[string]any:
		if len(typed) == 1 {
			if nestedType, ok := typed["string"]; ok {
				length := ctx.RandomGenerator.NextCollectionLength()
				example := make(map[string]any)
				for i := 0; i < length; i += 1 {
					example[ctx.RandomGenerator.NextString()] = generateRawTypeExample(nestedType, schema, ctx)
				}
				return example
			}
		}
	case map[any]any:
		return generateRawTypeExample(toStringAnyMap(typed), schema, ctx)
	}

	return nil
}

func normalizeExampleValue(value any) any {
	switch typed := value.(type) {
	case []byte:
		return base64.StdEncoding.EncodeToString(typed)
	case []any:
		normalized := make([]any, 0, len(typed))
		for _, entry := range typed {
			normalized = append(normalized, normalizeExampleValue(entry))
		}
		return normalized
	case map[string]any:
		normalized := make(map[string]any, len(typed))
		for key, entry := range typed {
			normalized[key] = normalizeExampleValue(entry)
		}
		return normalized
	case map[any]any:
		return normalizeExampleValue(toStringAnyMap(typed))
	default:
		return value
	}
}

func getDefaultFnScope(parsedTypes map[string]types.TType) string {
	nonInternalFunctions := make([]string, 0)
	allFunctions := make([]string, 0)

	for schemaKey := range parsedTypes {
		if !strings.HasPrefix(schemaKey, "fn.") || strings.HasSuffix(schemaKey, ".->") {
			continue
		}
		allFunctions = append(allFunctions, schemaKey)
		if !strings.HasSuffix(schemaKey, "_") {
			nonInternalFunctions = append(nonInternalFunctions, schemaKey)
		}
	}

	sort.Strings(nonInternalFunctions)
	if len(nonInternalFunctions) > 0 {
		return nonInternalFunctions[0]
	}

	sort.Strings(allFunctions)
	if len(allFunctions) > 0 {
		return allFunctions[0]
	}

	return "fn.ping_"
}

func getSchemaKey(definition map[string]any) string {
	for key := range definition {
		if key != "///" && key != "->" && key != "_errors" {
			return key
		}
	}

	panic(fmt.Errorf("telepact: schema entry has no schema key: %v", definition))
}
